### ISSUE TITLE
DI'ing the Container

### ISSUE THREAD

**Issue by jdahan91** [simpleinjector/SimpleInjector#805]
How can I DI the Container class? It does not implement an interface that is suitable to create Sdk's.

**Comment by dotnetjunkie** [comment]
Can you rephrase your question? What does it mean to "DI" the Container class? Are you referring to inject the `Container` into constructors of consumers, because that is supported out of the box. In case you want to let your classes depend on a (non-existing) `SimpleInjector.IContainer` interface instead of the concrete `SimpleInjector.Container` class, I would urge you to reconsider and read [this q&a](https://stackoverflow.com/questions/16225737/why-doesnt-simple-injector-have-an-icontainer-abstraction-like-unity).

As noted in that q&a I would suggest using common design practices and, for one, prevent letting low level code, reusable libraries and SDKs from having a dependency on *any* concrete DI implementation or even a Service Locator. Also take a look at Mark Seemann's articles on [DI-friendly libraries](https://blog.ploeh.dk/2014/05/19/di-friendly-library/) and [DI-friendly frameworks](https://blog.ploeh.dk/2014/05/19/di-friendly-framework/).

If you have any additional comments or questions based on my response, please let me know.

**Comment by jdahan91** [comment]
The thing is, I don't want to force developers to install SimpleInjector package just for declaring the "Container" part as not everyone needs this feature in their module ;].

So what can I do about this?

**Comment by dotnetjunkie** [comment]
I know too little about your OSS project to give very specific feedback on it, but I think that Mark Seemann's articles do a pretty good job in describing how to handle the general case.

**Comment by jdahan91** [comment]
Reading the QA from SoF link gives me a clue... I think the IServiceProvider is what i'm looking for ;]

**Comment by jdahan91** [comment]
I'm just trying to make my Sdk generic as much as possible, less dependencies for external developers.

**Comment by dotnetjunkie** [comment]
> I think the IServiceProvider is what i'm looking for ;]

Refrain from using `IServiceProvider` as it leads to the [Service Locator anti-pattern](https://mng.bz/WaQw).

> I'm just trying to make my Sdk generic as much as possible

In that case, consider defining special-purpose abstractions, such as factory abstractions (like `IControllerFactory`), in case you are writing *a framework* (as described [here](https://blog.ploeh.dk/2014/05/19/di-friendly-framework/) and consider using the builder pattern when writing a library (as described [here](https://blog.ploeh.dk/2014/05/19/di-friendly-framework/)). Factory abstractions are very explicit in the type they return and gives the consumer complete freedom over whether or not they want to use a DI container or not.

**Comment by jdahan91** [comment]
Let me tell you the real issue here, If i'm going to use a type defined in one of the SI library then the developer would be forced to include the package which not everyone need to use in their module developement.

**Comment by jdahan91** [comment]
> > I think the IServiceProvider is what i'm looking for ;]
> 
> Refrain from using `IServiceProvider` as it leads to the [Service Locator anti-pattern](https://mng.bz/WaQw).
> 
> > I'm just trying to make my Sdk generic as much as possible
> 
> In that case, consider defining special-purpose abstractions, such as factory abstractions (like `IControllerFactory`), in case you are writing _a framework_ (as described [here](https://blog.ploeh.dk/2014/05/19/di-friendly-framework/) and consider using the builder pattern when writing a library (as described [here](https://blog.ploeh.dk/2014/05/19/di-friendly-framework/)). Factory abstractions are very explicit in the type they return and gives the consumer complete freedom over whether or not they want to use a DI container or not.

Yes but creating this abstraction would still lead the Container type defined in the sdk (no?) and would cause the package to be installed within the module.

**Comment by dotnetjunkie** [comment]
> Yes but creating this abstraction would still lead the Container type defined in the sdk (no?) and would cause the package to be installed within the module.

You don't have to let the SDK depend on Simple Injector. If you don't want your SDK to depend on Simple Injector, you can provide users with an integration package. This allows you to define the factory abstraction(s) *in the SDK library* while having the factory implementation(s) (that depend on Simple Injector, or whatever DI Container you choose) in the integration library only.
